include/realtime_tools/realtime_box.h

```diff
@@ -26,52 +27,270 @@
+  }
+  /**
+   * @brief get the content and wait until the mutex could be locked
+   * @note same signature as in the existing RealtimeBox<T>
+   */
+  template <typename U = T>
+  typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, void> get(T & in) const
+  {
+    std::lock_guard<mutex_t> guard(lock_);
+    // cppcheck-suppress missingReturn
+    in = value_;
+  }
+  /**
+   * @brief access the content (r) and wait until the mutex could be locked
+   * @note only safe way to access pointer type content (r)
+   * @note same signature as in the existing RealtimeBox<T>
+   */
+  void get(const std::function<void(const T &)> & func)
+  {
+    std::lock_guard<mutex_t> guard(lock_);
+    func(value_);
+  }
+
+  /**
+   * @brief provide a custom assignment operator for easier usage
+   * @note only to be used from non-RT!
+   */
+  template <typename U = T>
+  typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, void> operator=(const T & value)
+  {
+    set(value);
+  }
+
+  /**
+   * @brief provide a custom conversion operator
+   * @note Can only be used from non-RT!
+   */
+  template <typename U = T, typename = typename std::enable_if_t<!is_ptr_or_smart_ptr<U>>>
+  [[nodiscard]] operator T() const
+  {
+    // Only makes sense with the getNonRT method otherwise we would return an std::optional
+    return get();
+  }
+  /**
+   * @brief provide a custom conversion operator
+   * @note Can be used from non-RT and RT contexts
+   */
+  template <typename U = T, typename = typename std::enable_if_t<!is_ptr_or_smart_ptr<U>>>
+  [[nodiscard]] operator std::optional<T>() const
+  {
+    return try_get();
+  }
+
+  // In case one wants to actually use a pointer
+  // in this implementation we allow accessing the lock directly.
+  // Note: Be careful with lock.unlock().
+  // It may only be called from the thread that locked the mutex!
+  [[nodiscard]] const mutex_t & get_mutex() const { return lock_; }
+  [[nodiscard]] mutex_t & get_mutex() { return lock_; }
 
 private:
-  // The thing that's in the box.
-  T thing_;
+  T value_;
 
   // Protects access to the thing in the box.  This mutex is
   // guaranteed to be locked for no longer than the duration of the
   // copy, so as long as the copy is realtime safe and the OS has
   // priority inheritance for mutexes, this lock can be safely locked
   // from within realtime.
-  std::mutex thing_lock_RT_;
+  mutable mutex_t lock_;
 };
 
+// Introduce some easier to use names
+
+// Only kept for compatibility reasons
+template <typename T, typename mutex_type = std::mutex>
+using RealtimeBoxBestEffort [[deprecated]] = RealtimeBox<T, mutex_type>;
+
+// Provide specialisations for different mutex types
+template <typename T>
+using RealtimeBoxStandard = RealtimeBox<T, std::mutex>;
+
+template <typename T>
+using RealtimeBoxRecursive = RealtimeBox<T, std::recursive_mutex>;
+
+// This is the specialisation we recommend to use in the end
+template <typename T>
+using RealtimeBox = RealtimeBoxDefault<T>;
```

Comment:
Which is the default one? I don't see it's definition in this file

Reply:
Sorry I just messed up a little bit. Can you take a look at the latest commit ?

Reply:
Sure

Reply:
Looks good to me